File: docs/test-cases/insecure-plugin-design.java
// Test case: insecure-plugin-design (LLM07:2025)
package com.example.tools;

import dev.langchain4j.agent.tool.Tool;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;

public class LlmTools {

    private final HttpClient httpClient = HttpClients.createDefault();

    // BUG: @Tool string parameter has no schema constraints (no maxLength, no pattern).
    // BUG: path is read directly with no confinement to an allowed base directory —
    // LLM can request "../../etc/passwd" or "/home/user/.ssh/id_rsa".
    @Tool("Read a file from disk")
    public String readFile(String path) throws Exception {
        return Files.readString(Path.of(path));
    }

    // BUG: raw SQL from the model is concatenated into a JDBC Statement — the tool
    // description says "query" but the LLM can emit DROP TABLE, UPDATE, or UNION
    // SELECT against other tables. No PreparedStatement, no allowlist of tables.
    @Tool("Run a SQL query against the app database")
    public String runQuery(String query) throws Exception {
        try (Connection c = DriverManager.getConnection("jdbc:postgresql://db/app");
             Statement stmt = c.createStatement();
             ResultSet rs = stmt.executeQuery(query)) {
            StringBuilder sb = new StringBuilder();
            while (rs.next()) sb.append(rs.getString(1)).append("\n");
            return sb.toString();
        }
    }

    // BUG: no URL scheme/host allowlist — the LLM can be coerced via prompt injection
    // into fetching http://169.254.169.254/latest/meta-data/ (SSRF to cloud metadata)
    // or file:// URLs. No authorization check on the caller either.
    @Tool("Fetch a URL and return the body")
    public String fetchUrl(String url) throws Exception {
        return EntityUtils.toString(httpClient.execute(new HttpGet(url)).getEntity());
    }
}

File: docs/test-cases/insecure-plugin-design.rs
// Test case: insecure-plugin-design (LLM07:2025)
use rusqlite::Connection;
use serde_json::json;
use std::fs;

// Tool schemas handed to the Anthropic API. Every string parameter is bare —
// no maxLength, no pattern, no enum — so the model can pass anything.
pub fn tool_definitions() -> serde_json::Value {
    json!([
        {
            "name": "read_file",
            "description": "Read a file from disk",
            // BUG: unconstrained string parameter
            "input_schema": {
                "type": "object",
                "properties": { "path": { "type": "string" } },
                "required": ["path"]
            }
        },
        {
            "name": "run_query",
            "description": "Run a SQL query",
            // BUG: unconstrained string parameter, full SQL surface exposed
            "input_schema": {
                "type": "object",
                "properties": { "sql": { "type": "string" } },
                "required": ["sql"]
            }
        },
        {
            "name": "fetch",
            "description": "Fetch a URL",
            // BUG: unconstrained string parameter, no scheme allowlist
            "input_schema": {
                "type": "object",
                "properties": { "url": { "type": "string" } },
                "required": ["url"]
            }
        }
    ])
}

// BUG: no canonicalization, no base-directory confinement — LLM can request
// "../../etc/shadow" or "/root/.aws/credentials".
pub fn read_file(path: String) -> std::io::Result<String> {
    fs::read_to_string(path)
}

// BUG: raw SQL from the model is executed — no parameterization, no table
// allowlist, DROP/UPDATE/ATTACH all reachable.
pub fn run_query(sql: String) -> rusqlite::Result<()> {
    let conn = Connection::open("app.db")?;
    conn.execute(&sql, [])?;
    Ok(())
}

// BUG: no scheme/host allowlist — reqwest will happily follow file://,
// http://169.254.169.254/ (cloud metadata SSRF), or internal RFC1918 targets.
pub fn fetch(url: String) -> Result<String, reqwest::Error> {
    reqwest::blocking::get(&url)?.text()
}
